### ISSUE TITLE
Multiple results should generate valid JSON output

### ISSUE THREAD

**Issue by uvsmtid** [saltstack/salt#25153]
Based on [this stackoverflow question][1].

## Problem ##

Command:
```
salt --out json '*' test.ping
```

Output (for 2 minions):
```
{
    "monkey": true
}
{
    "elephant": true
}
```
This output is not [valid JSON][2] which cannot contain more than one top-level objects (begins with `{` and ends with `}`).

The problem is that instead of using JSON output directly, one has to create a [special parser][3] to recognize the beginning and the end of multiple JSON objects, split the objects, feed them one by one to the parser. Or, in short, the output is unusable.

## Test ##

`load_json.py`:
```
import sys
import json
data = json.load(sys.stdin)
print str(data)
```

For example, capturing the output above and feeding it into standard JSON parser in Python does not work:
```
salt --out json '*' test.ping | python load_json.py
Traceback (most recent call last):
  File "load_json.py", line 4, in <module>
    data = json.load(sys.stdin)
  File "/usr/lib64/python2.7/json/__init__.py", line 290, in load
    **kw)
  File "/usr/lib64/python2.7/json/__init__.py", line 338, in loads
    return _default_decoder.decode(s)
  File "/usr/lib64/python2.7/json/decoder.py", line 369, in decode
    raise ValueError(errmsg("Extra data", s, end, len(s)))
ValueError: Extra data: line 4 column 1 - line 7 column 1 (char 23 - 48)
```

## Versions ##

```
 salt --versions
           Salt: 2015.5.2
         Python: 2.7.10 (default, May 27 2015, 18:11:38)
         Jinja2: 2.7.3
       M2Crypto: 0.21.1
 msgpack-python: 0.4.6
   msgpack-pure: Not Installed
       pycrypto: 2.6.1
        libnacl: Not Installed
         PyYAML: 3.11
          ioflo: Not Installed
          PyZMQ: 14.4.1
           RAET: Not Installed
            ZMQ: 4.0.5
           Mako: 1.0.1
```

[1]: http://stackoverflow.com/q/29960947/441652
[2]: http://json.org/
[3]: https://github.com/uvsmtid/common-salt-states/blob/75e42afb04cbd7d82722c9793c1c9be6637b8bbd/states/bootstrap/bootstrap.dir/modules/utils/check_salt_output.py#L107

**Comment by uvsmtid** [comment]
#14725 is a similar issue where output turns unusable due to other data on STDOUT.

**Comment by uvsmtid** [comment]
#25153 is a similar issue to make sure all features together provide valid output. 
#14725 is a similar issue specific to salt-cloud.

**Comment by uvsmtid** [comment]
AFAIK, multiple results are independently generated by multiple minions (and each run of function on specific minion actually produces valid output).

What could be done on the `salt` command side (run on the master) is to consolidate data from all minions into internal in-memory object and render it into JSON or any other format after (rather than sending JSON to STDOUT without consolidation). To avoid scalability issues, at least it should appear as if it is done so.

Or, particularly in this case, there is a hack to wrap individual output into JSON array:
```
[
...
,
...
,
...
]
```
The problem is that, in order to avoid parsing by user, it still can only be done on master at least to inject commas `,`. And this will only work for JSON, and will only look proper when results are of the same type (conceptually, array contains objects of the same type, semantic, structure, etc.).

**Comment by uvsmtid** [comment]
I guess this issue is supposed to be fixed by providing `--static` option and [this explanation in documentation][1].

At the moment (`2015.5.2`) my tests fail due to fixed issue #24776 which I'll try as soon as native Fedora package is updated.

[1]: https://github.com/saltstack/salt/blob/561d7b72bb8bb0d5fa5c4395db12425f27c26c1a/doc/man/salt.1#L282-L285

**Comment by uvsmtid** [comment]
The issue can be considered fixed in `2015.5.3`.

### Test ###

Without `--static`:
```
salt --out json '*' test.ping
jid: 20150717013306606937
{
    "projector": true
}
{
    "rhel7_minion": true
}
{
    "rhel5_minion": "Minion did not return. [No response]"
}
```

With `--static`:
```
 salt --out json --static '*' test.ping
{
    "rhel7_minion": true, 
    "projector": true
}
```

### Results ###

*   With respect to this issue, the `--static` output for multiple minions is definitely a valid JSON.
*   The result should also close #25154.
*   One can argue that non-returning minion (see `rhel5_minion` above) should be included in the `--static` output with result `false`, but this is a separate topic. [This comment on another issue][1] indicates direction:
  _... I think the pattern is swinging in the direction of "minions which didn't return are treated as an error"_


[1]: https://github.com/saltstack/salt/issues/13216#issuecomment-88974415

**Comment by rallytime** [comment]
@uvsmtid Thanks very much for this excellent analysis and for the pull request. Since you've stated above that this is considered to be fixed by you in 2015.5.3, I am going to go ahead and close this issue. Thanks again!
